File: app/src/main/java/com/tokyonth/txphook/utils/json/HookConfigManager.java
package com.tokyonth.txphook.utils.json;

import com.tokyonth.txphook.Constants;
import com.tokyonth.txphook.entry.HookAppsEntry;
import com.tokyonth.txphook.entry.HookInfoEntry;
import com.tokyonth.txphook.utils.file.FileUtils;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

public class HookConfigManager {

    private static HookConfigManager instance;

    public static HookConfigManager getInstance() {
        if (instance == null) {
            instance = new HookConfigManager();
        }
        return instance;
    }

    public ArrayList<HookAppsEntry> getAppConfig() {
        File jsonFile = new File(Constants.HOOK_JSON_PATH + Constants.HOOK_APPS_CONFIG);

        String appsConfig = FileUtils.read(jsonFile);
        ArrayList<HookAppsEntry> configList = new ArrayList<>();
        if (!appsConfig.isEmpty()) {
            try {
                JSONArray jsonArray = new JSONArray(appsConfig);
                for (int i = 0; i < jsonArray.length(); i++) {
                    JSONObject jsonObject = jsonArray.getJSONObject(i);
                    HookAppsEntry hookApps = new HookAppsEntry();
                    hookApps.packageName = jsonObject.getString("packageName");
                    hookApps.isHook = jsonObject.getBoolean("isHook");
                    configList.add(hookApps);
                }
            } catch (JSONException e) {
                e.printStackTrace();
            }
        }
        return configList;
    }

    public void saveAppConfig(HookAppsEntry appsEntry) {
        String jsonPath = Constants.HOOK_JSON_PATH + Constants.HOOK_APPS_CONFIG;
        String appsConfig = FileUtils.read(new File(jsonPath));

        try {
            JSONArray jsonArray;
            if (appsConfig.isEmpty()) {
                jsonArray = new JSONArray();
            } else {
                jsonArray = new JSONArray(appsConfig);
            }

            JSONObject jsonObject = new JSONObject();
            jsonObject.put("packageName", appsEntry.packageName);
            jsonObject.put("isHook", appsEntry.isHook);
            jsonArray.put(jsonObject);
            FileUtils.write(jsonArray.toString(), Constants.HOOK_JSON_PATH, Constants.HOOK_APPS_CONFIG);
        } catch (JSONException e) {
            e.printStackTrace();
        }
    }

    public ArrayList<HookInfoEntry> getHookConfig(String pkgName) {
        String jsonPath = Constants.HOOK_JSON_PATH + pkgName + ".json";
        String appsConfig = FileUtils.read(new File(jsonPath));
        ArrayList<HookInfoEntry> configList = new ArrayList<>();
        if (!appsConfig.isEmpty()) {
            try {
                JSONArray jsonArray = new JSONArray(appsConfig);
                for (int i = 0; i < jsonArray.length(); i++) {
                    JSONObject jsonObject = jsonArray.getJSONObject(i);
                    HookInfoEntry config = new HookInfoEntry();
                    config.packageName = jsonObject.getString("packageName");
                    config.hookName = jsonObject.getString("hookName");
                    config.classPath = jsonObject.getString("classPath");
                    config.methodName = jsonObject.getString("methodName");
                    config.resultVale = jsonObject.get("resultVale");
                    configList.add(config);
                }
            } catch (JSONException e) {
                e.printStackTrace();
            }
        }
        return configList;
    }

    public void removeHookConfig(String pkgName, int index) {
        String jsonPath = Constants.HOOK_JSON_PATH + pkgName + ".json";
        String appsConfig = FileUtils.read(new File(jsonPath));
        if (appsConfig.isEmpty())
            return;
        try {
            JSONArray jsonArray = new JSONArray(appsConfig);
            jsonArray.remove(index);
            String fileName = pkgName + ".json";
            FileUtils.write(jsonArray.toString(), Constants.HOOK_JSON_PATH, fileName);
        } catch (JSONException e) {
            e.printStackTrace();
        }
    }

    public void saveHookConfig(HookInfoEntry hookInfoEntry) {
        String jsonPath = Constants.HOOK_JSON_PATH + hookInfoEntry.packageName + ".json";
        String appsConfig = FileUtils.read(new File(jsonPath));

        try {
            JSONArray jsonArray;
            if (appsConfig.isEmpty()) {
                jsonArray = new JSONArray();
            } else {
                jsonArray = new JSONArray(appsConfig);
            }

            JSONObject jsonObject = new JSONObject();
            jsonObject.put("packageName", hookInfoEntry.packageName);
            jsonObject.put("hookName", hookInfoEntry.hookName);
            jsonObject.put("classPath", hookInfoEntry.classPath);
            jsonObject.put("methodName", hookInfoEntry.methodName);
            jsonObject.put("resultVale", hookInfoEntry.resultVale);
            jsonArray.put(jsonObject);
            String fileName = hookInfoEntry.packageName + ".json";
            FileUtils.write(jsonArray.toString(), Constants.HOOK_JSON_PATH, fileName);
        } catch (JSONException e) {
            e.printStackTrace();
        }
    }

    public void saveAllHookConfigs(List<HookInfoEntry> hookInfoEntries) {
        try {
            JSONArray jsonArray = new JSONArray();
            for (HookInfoEntry hookInfoEntry : hookInfoEntries) {
                JSONObject jsonObject = new JSONObject();
                jsonObject.put("packageName", hookInfoEntry.packageName);
                jsonObject.put("hookName", hookInfoEntry.hookName);
                jsonObject.put("classPath", hookInfoEntry.classPath);
                jsonObject.put("methodName", hookInfoEntry.methodName);
                jsonObject.put("resultVale", hookInfoEntry.resultVale);
                jsonArray.put(jsonObject);
            }
            String fileName = hookInfoEntries.get(0).packageName + ".json";
            FileUtils.write(jsonArray.toString(), Constants.HOOK_JSON_PATH, fileName);
        } catch (JSONException e) {
            e.printStackTrace();
        }
    }

}

File: app/src/main/java/com/tokyonth/txphook/utils/json/HookConfigManager.kt
package com.tokyonth.txphook.utils.json

import android.Manifest
import androidx.appcompat.app.AppCompatActivity
import com.tokyonth.txphook.Constants
import com.tokyonth.txphook.db.HookAppInfo
import com.tokyonth.txphook.db.HookConfig
import com.tokyonth.txphook.db.HookRule
import com.tokyonth.txphook.utils.file.FileUtils
import com.tokyonth.txphook.utils.permission.PermissionUtils
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import java.io.File

class HookConfigManager {

    companion object {

        private var hookConfigManager: HookConfigManager? = null

        private val permissionArray = arrayOf(
            Manifest.permission.READ_EXTERNAL_STORAGE,
            Manifest.permission.READ_EXTERNAL_STORAGE
        )

        fun of(activity: AppCompatActivity, instance: (HookConfigManager?) -> Unit) {
            PermissionUtils(activity)
                .request(*permissionArray) { areGrantedAll, _ ->
                    if (areGrantedAll) {
                        if (hookConfigManager == null) {
                            hookConfigManager = HookConfigManager()
                        }
                        instance.invoke(hookConfigManager)
                    } else {
                        instance.invoke(null)
                    }
                }
        }

    }

    fun export(hookAppInfo: HookAppInfo): Boolean {
        val config = JSONObject().apply {
            put("appName", hookAppInfo.config.appName)
            put("appVersion", hookAppInfo.config.appVersion)
            put("packageName", hookAppInfo.config.packageName)
        }

        val rules = JSONArray()
        hookAppInfo.rule.forEach {
            val rule = JSONObject().apply {
                put("enableHook", it.enableHook)
                put("pkgName", it.pkgName)
                put("hookName", it.hookName)
                put("classPath", it.classPath)
                put("methodName", it.methodName)
                put("resultVale", it.resultVale)
                put("valueType", it.valueType)
            }
            rules.put(rule)
        }
        val result = JSONObject().apply {
            put("config", config)
            put("rules", rules)
        }

        val fileName = hookAppInfo.config.packageName + ".json"
        FileUtils.write(result.toString(), Constants.HOOK_JSON_PATH, fileName)
        return true
    }

    fun import(filePath: String, success: (HookAppInfo) -> Unit, error: (String) -> Unit) {
        val jsonContent = FileUtils.read(File(filePath))
        if (jsonContent.isEmpty()) {
            error.invoke("配置为空!")
            return
        }
        try {
            val json = JSONObject(jsonContent)
            val config = json.getJSONObject("config")
            val rules = json.getJSONArray("rules")

            val pkgName = config.getString("packageName")
            val appName = config.getString("appName")
            val appVersion = config.getString("appVersion")
            val configEntity = HookConfig(0, appName, pkgName, appVersion)

            val resultRule: MutableList<HookRule> = ArrayList()
            for (index in 0 until rules.length()) {
                val obj = rules.getJSONObject(index)
                val enableHook = obj.getBoolean("enableHook")
                val pkgNameR = obj.getString("pkgName")
                val hookName = obj.getString("hookName")
                val classPath = obj.getString("classPath")
                val methodName = obj.getString("methodName")
                val resultVale = obj.getString("resultVale")
                val valueType = obj.getInt("valueType")

                val rule = HookRule(
                    0,
                    enableHook,
                    pkgNameR,
                    hookName,
                    classPath,
                    methodName,
                    resultVale,
                    valueType
                )
                resultRule.add(rule)
            }
            val hookAppInfo = HookAppInfo(configEntity, resultRule)
            success.invoke(hookAppInfo)
        } catch (e: JSONException) {
            error.invoke("配置解析错误!")
        }
    }

}
